.tools/validate.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+var (
+	flCommitRange = flag.String("range", "", "use this commit range instead")
+
+	validDCO = regexp.MustCompile(`^Signed-off-by: ([^<]+) <([^<>@]+@[^<>]+)>$`)
+)
+
+var DefaultRules = []ValidateRule{
+	func(c CommitEntry) error {
+		hasValid := false
+		for _, line := range strings.Split(c.Body, "\n") {
+			if validDCO.MatchString(line) {
+				hasValid = true
+			}
+		}
+		if !hasValid {
+			return fmt.Errorf("%q does not have a valid DCO", c.Commit)
+		}
+		return nil
+	},
+	// TODO add something for the cleanliness of the c.Subject
+}
+
+func main() {
+	flag.Parse()
+
+	/*
+		if len(os.Getenv("TRAVIS")) > 0 {
+			fmt.Println(os.Getenv("TRAVIS_COMMIT_RANGE"))
+			if len(os.Getenv("TRAVIS_PULL_REQUEST")) > 0 {
+				fmt.Println(os.Getenv("TRAVIS_BRANCH"))
+			}
+		}
+	*/
+
+	invalids := []InvalidCommit{}
+	if *flCommitRange != "" {
+		c, err := GitCommits(*flCommitRange)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		for _, commit := range c {
+			fmt.Printf(" * %s %s ... ", commit.AbbreviatedCommit, commit.Subject)
+			if err := ValidateCommit(commit, DefaultRules); err != nil {
+				invalids = append(invalids, InvalidCommit{commit, err})
+				fmt.Printf("FAIL: %s\n", err)
+			} else {
+				fmt.Println("PASS")
+			}
+		}
+	} else if os.Getenv("TRAVIS_COMMIT_RANGE") != "" {
+		c, err := GitCommits(os.Getenv("TRAVIS_COMMIT_RANGE"))
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		for _, commit := range c {
+			fmt.Printf(" * %s %s ... ", commit.AbbreviatedCommit, commit.Subject)
+			if err := ValidateCommit(commit, DefaultRules); err != nil {
+				invalids = append(invalids, InvalidCommit{commit, err})
+				fmt.Printf("FAIL: %s\n", err)
+			} else {
+				fmt.Println("PASS")
+			}
+		}
+	}
+	if len(invalids) > 0 {
+		fmt.Printf("%d issues to fix\n", len(invalids))
+		os.Exit(1)
+	}
+}
+
+type ValidateRule func(CommitEntry) error
+
+func ValidateCommit(c CommitEntry, rules []ValidateRule) error {
+	for _, r := range rules {
+		if err := r(c); err != nil {
+			return err
```

Comment:
On Wed, Sep 09, 2015 at 12:50:47PM -0700, Vincent Batts wrote:
> perhaps. Though there is only the one test for now.

Sure.  I'm ok punting until we add another test, although I'm not sure
what my odds are of remembering to comment again then ;).  An easier
approach would be to just log the error messages directly (maybe with
TAP-friendly output?) from this helper, and then return a boolean for
“there were (or weren't) errors with this commit”.